---
title: "FMR1 Knockout"
subtitle: "RNA-Seq Analysis"
author: "Melanie Hand & Steve Pederson"
date: "19/07/2019"
output: html_document
---

## Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r packages}
library(limma)
library(edgeR)
library(AnnotationHub)
library(tidyverse)
library(magrittr)
library(pander)
library(RColorBrewer)
library(ggrepel)
library(scales)
library(here)
theme_set(theme_bw())
panderOptions("big.mark", ",")
panderOptions("table.split.table", Inf)
panderOptions("table.style", "rmarkdown")
if (interactive()) setwd(file.path(here(), "R"))
```

Annotation was setup as a EnsDb object based on Ensembl Release 94

```{r ensDb, cache=TRUE}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(dataprovider == "Ensembl") %>%
  subset(rdataclass == "EnsDb")
ensDb <- AnnotationHub()[["AH64906"]]
genes <- genes(ensDb)
```

Prior to count-level analysis, the initial dataset was pre-processed using the following steps:

- Adapters were removed from any reads derived from RNA fragments < 300bp
- Bases were removed from the end of reads when the quality score dipped below 20
- Reads < 35bp after trimming were discarded

After trimming alignment was performed using STAR v2.5.3a to the *Danio rerio* genome included in Ensembl Release 94 (GRCz11).
Aligned reads were counted for each gene if the following criteria were satisfied:

- Alignments were unique
- Alignments stricly overlapped exonic regions

Counts were then loaded and setup as a DGEList.
During loading genes were only retained if receiving more than one counted alignment in at least 4 samples.


```{r x}
x <- file.path("../2_alignedData/featureCounts/genes.out") %>%
	read_tsv() %>%
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove(colnames(.), "_Aligned.+")) %>%
	magrittr::extract(rowSums(cpm(.) > 1) >= 4,) %>%
	DGEList(
		samples = tibble(
			sample = colnames(.),
			group = case_when(
                    sample %in% paste0("S", 1:8) ~ "FMR1",
                    !sample %in% paste0("S", 1:8)  ~ "Wildtype"
                  )) %>%
			mutate(group = factor(group, levels = c("Wildtype", "FMR1"))) %>%
			as.data.frame(),
		genes = genes[rownames(.)] %>%
			as.data.frame() %>%
			dplyr::select(
				ensembl_gene_id = gene_id,
				chromosome_name = seqnames,
				description, 
				external_gene_name = gene_name
			)
	) %>%
	calcNormFactors(method = "TMM") %>%
	estimateDisp()
```

This gave count information for `r nrow(x)` genes.
Library sizes after alignment, counting and gene filtering ranged between `r pander(comma(range(x$samples$lib.size)))` reads.

```{r plotCols}
nsamples <- ncol(x)
col <- brewer.pal(nsamples, "Set1")
```

## Data Inspection

Data was subsequently configured for analysis using voom, in which the assumption of normality can be applied and sample weights cn be incorporated.
Sample weights were calculated assuming all samples belong to the same treatment group to ensure conservative weights are assigned.

```{r v}
v <- x %>%
	voomWithQualityWeights(design = matrix(1, nrow = ncol(.)))
```

```{r plotWeights, echo=FALSE, fig.cap = "Sample weights with the ideal equal weight of 1 shown as a horizontal line."}
v$targets %>%
	ggplot(aes(sample, sample.weights, fill = group)) +
	geom_bar(stat = "identity") +
	geom_hline(yintercept = 1, colour = "blue", linetype = 2) +
	labs(fill = "Sample Group")
```


```{r pca}
pca <- v$E %>%
	t() %>%
	prcomp()
```

A PCA was performed using logCPM transformed counts after removal of undetectable genes.
Whilst some separation was observed, the location of sample S5 appeared less than satisfactory, and the spread between Wildtype samples also appeared to coincide with the largest source of variability in the data.

```{r plotPCA, echo=FALSE, fig.cap = "PCA of all samples. Point sizes indicate library sizes with the combination of PC1 & PC2 appearing to capture a considerable amount of this variation."}
pca$x %>%
	as.data.frame() %>%
	rownames_to_column("sample") %>%
	left_join(v$targets) %>%
	ggplot(aes(PC1, PC2, colour = group)) +
	geom_point(aes(size = lib.size/1e6)) +
	geom_text_repel(aes(label = sample), show.legend = FALSE) +
	labs(size = "Library Size\n(millions)",
		 colour = "Sample Group") +
	theme(legend.title.align = 0.5)
```

Finally, genotypes were checked using *fmr1* expression as a proxy.
Expression of *fmr1* appeared to track with genotype as expected and no sample mislabelling was identified.

```{r plotFMR1, echo=FALSE, fig.cap = "Expression of *fmr1* across all samples."}
x %>%
	cpm(log = TRUE) %>%
	.["ENSDARG00000037433",] %>%
	as.matrix() %>%
	set_colnames("logCPM") %>%
	cbind(x$samples[rownames(.),]) %>%
	ggplot(aes(sample, logCPM, fill = group)) +
	geom_bar(stat = "identity")
```


## Analysis

The `voom` methodology was applied for differential gene expression (DGE) analysis, incorporating sample weights as calculated earlier.
A simple two group design matrix was used.

```{r efit}
design <- model.matrix(~v$targets$group)
colnames(design) <- c("Wildtype", "FMR1vsWildtype")
efit <- v %>%
	lmFit(design) %>%
	eBayes()
```

```{r deGenes}
deGenes <- efit %>%
	topTable(coef = "FMR1vsWildtype", number = Inf, sort.by = "p") %>%
	as_tibble()
```

```{r volcano, echo=FALSE, fig.cap = "Volcano plot highlighting DE genes"}
deGenes %>%
	mutate(Sig = adj.P.Val < 0.05) %>%
	ggplot(aes(logFC, -log10(P.Value), colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = external_gene_name), 
					data = . %>% dplyr::filter(Sig)) +
	geom_vline(xintercept = c(-1, 1), linetype = 2, colour = "grey50") +
	geom_hline(yintercept = deGenes %>%
			   	dplyr::filter(adj.P.Val < 0.05) %>%
			   	.[["P.Value"]] %>%
			   	max() %>% 
			   	log(10) %>%
			   	multiply_by(-1), 
			   linetype = 2, colour = "blue") +
	scale_colour_manual(values = c(rgb(0.5, 0.5, 0.5, 0.5), "red")) +
	scale_x_continuous(breaks = seq(-6, 6, by = 2)) +
	theme(legend.position = "none")
```

```{r plotMA, echo=FALSE, fig.cap="logFC plotted against expression level with significant DE genes shown in red."}
deGenes %>%
mutate(Sig = adj.P.Val < 0.05) %>%
	ggplot(aes(AveExpr, logFC, colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = external_gene_name), 
					data = . %>% dplyr::filter(Sig | abs(logFC) > 4.5)) +
	geom_hline(yintercept = c(-1, 1), linetype = 2, colour = "grey50") +
	labs(
		x = "Average Expression (logCPM)",
		y = "logFC"
	) +
	scale_colour_manual(values = c(rgb(0.5, 0.5, 0.5, 0.5), "red")) +
	theme(legend.position = "none")
```

