---
title: "FMR1 Analysis Using GLM Methodologies"
author: "Karissa Barthelson, Melanie Hand and Stephen Pederson"
date: "12/05/2020"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r packages}
library(magrittr)
library(tidyverse)
library(limma)
library(edgeR)
library(AnnotationHub)
library(fgsea)
library(msigdbr)
library(goseq)
library(scales)
library(DT)
library(ggrepel)
library(ggeasy)
library(cqn)
library(ggfortify)
library(cowplot)
library(pathview)
library(pander)

library(fgsea)
library(biomaRt)
library(Gviz)
library(BSgenome.Drerio.UCSC.danRer11)

theme_set(theme_bw())
```


# Introduction

This analysis will be doing a differential gene expression and differential pathway analysis from *fmr1* knockout larvae at 2dpf. [Previously](FMR1_Analysis.html), Stephen Pederson and Melanie Hand did a DE analysis using *limma*.
Here, I will do one with the generalised linear model (GLM) capablities of edgeR as this, in my experience, detects more DE genes. 
I then will do some pathway analysis on DE genes (if we get enough), and also the entire dataset using ranked lists. 

## Data description

This was taken from *FMR1_analysis.Rmd*

Prior to count-level analysis, the initial dataset was pre-processed using the following steps:

- Adapters were removed from any reads derived from RNA fragments < 300bp
- Bases were removed from the end of reads when the quality score dipped below 20
- Reads < 35bp after trimming were discarded

After trimming alignment was performed using STAR v2.5.3a to the *Danio rerio* genome included in Ensembl Release 94 (GRCz11).
Aligned reads were counted for each gene if the following criteria were satisfied:

- Alignments were unique
- Alignments strictly overlapped exonic regions

Counts were then loaded and set up as a DGEList.
During loading genes were only retained if receiving more than one counted alignment in at least 4 samples.

## Make annotation object and read in gene counts 

```{r ensDb, cache=TRUE}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(dataprovider == "Ensembl") %>%
  subset(rdataclass == "EnsDb")
ensDb <- AnnotationHub()[["AH64906"]]
genes <- genes(ensDb)
```

```{r}
x <- here::here("2_alignedData/featureCounts/genes.out") %>%
	read_tsv() %>%
	as.data.frame() %>%
	column_to_rownames("Geneid") %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove(colnames(.), "_Aligned.+")) %>%
	magrittr::extract(rowSums(cpm(.) > 1) >= 4,) %>%
	DGEList(
		samples = tibble(
			sample = colnames(.),
			group = case_when(
                    sample %in% paste0("S", 1:8) ~ "FMR1",
                    !sample %in% paste0("S", 1:8)  ~ "Wildtype"
                  )) %>%
			mutate(group = factor(group, levels = c("Wildtype", "FMR1"))) %>%
			as.data.frame(),
		genes = genes[rownames(.)] %>%
			as.data.frame() %>%
			mutate(location = paste0(seqnames, ":", start, "-", end, ":", strand)) %>%
			dplyr::select(
				ensembl_gene_id = gene_id,
				location,
				description, 
				external_gene_name = gene_name
			)
	) %>%
	calcNormFactors(method = "TMM") %>%
	estimateDisp()
```

# PCA analysis:

The overall similarity between samples can be explored by principal component analysis (PCA). PCA is a dimension reducing technique which uses linear combinations of the original data (i.e. the logCPMs of the genes in the RNA-seq dataset) to define a new set of variables (principal components).

Samples which have similar gene expression profiles will cluster together in a plot of principal components (PC) PC1 vs PC2. 
In the graph below, we see some seperation of samples by genotype across PC1. 
This means that *fmr1* genotype is the largest source of variation and explains ~35% of the variance. 
Sample S5 doesnt group with the rest of the *fmr1* samples, so it is likely an outlier. 
I will do a DE and pathway analysis including it. 

```{r, fig.cap = "*PCA analysis using logCPM values after filtering of undectable genes. Sample S5 does appear most similar to the wiltype samples, however the genotype has been checked via expression of fmr1*"}
pca_logCPM <- x$counts %>% 
	cpm(log = TRUE) %>% 
	t() %>% 
	prcomp()
pca_logCPM %>%
	autoplot(data = x$samples, colour = "group", size = 4) +
	geom_text_repel(
		aes(colour = group, label = sample), 
		show.legend = FALSE
	) +
	labs(colour = "Genotype")
```

# DE Analysis  

I will use the generalised linear model functionality of the package *edgeR* to perform the DE analysis. edgeR uses a negative binomial variance function and estimates dispersions using the Cox-Reid profile-adjusted likelihood
(CR) method. Tests for differential expression were perforned using likelihood ratio tests. 

## Design Matrix

The design matrix specifies which samples contain which characteristics. Here, WT was set as the intercept (or baseline) and the effect of FMR1 KO was set as the slope coefficient. 

```{r design}
design <- model.matrix(~group, data = x$samples)
colnames(design) <- str_remove(colnames(design), "group")
```

## Model Fit
Dispersions were already esitmated when generating the dge object *x*. This code below fits the negative binomial model. 

```{r}
glmFit <- glmFit(x, design)
```

## Likelihood ratio tests

This code below tests for differential expression. 
Genes were considered DE if the FDR adjusted p-value was < 0.05. 

```{r}
topTable_glm <- glmFit %>% 
	glmLRT(coef = "FMR1") %>%
	topTags(n = Inf, adjust.method = "fdr", sort.by = "p") %>% 
	.$table %>% 
	as_tibble() %>% 
	mutate(DE = FDR < 0.05)
```

## MD plot

Mean difference plots (MD)-plots show the average expression vs logFC. 
The DE genes had low/mild average expression (logCPM between ~0.3 and 6.7). 
Importantly, there appears to be a small bias in this dataset.
This may impact GSEA analysis as genesets with low-medium expressed genes will appear as enriched for being upregulated.

```{r}
md1 <- topTable_glm %>% 
	ggplot(aes(x = logCPM, y = logFC)) +
	geom_point(aes(colour = DE), alpha = 0.5) +
	geom_smooth(se = FALSE) +
	scale_color_manual(values = c("grey50", "red")) +
	geom_label_repel(
		aes(label = external_gene_name, colour = DE), 
		data = . %>% dplyr::filter(DE == TRUE), 
		show.legend = FALSE
	) 
md1
```


## CQN

In response to the observed bias CQN was used to rectify this issue as it may be derived from either a length or GC artefact.
GC and length information is easily avaaible from `EnsDb` objects from release 98 onwards, and this release was used.
No expresed genes were noted as absent from the annotations for Release 98, despite the differences between releases.
GC content and Length were taken as weighted averages and simple averages respectively.

```{r}
ens98 <- ah[["AH74989"]]
ex <- exonsBy(ens98, "tx")
tr <- transcripts(ens98)
tr$len <- ex %>%
	width %>%
	lapply(sum) %>%
	.[names(tr)] %>%
	unlist()
gnBias <- mcols(tr) %>%
	as.data.frame() %>%
	group_by(gene_id) %>%
	summarise(
		n = n(),
		meanGC = sum(len*gc_content) / sum(len),
		len = mean(len)
	)
mcols(genes) %<>%
	as.data.frame() %>%
	left_join(gnBias) %>%
	DataFrame()
```

```{r}
gcCqn <- cqn(
	counts = x$counts,
	x = enframe(rownames(x), name = c()) %>%
		dplyr::rename(gene_id = value) %>% 
		left_join(gnBias) %>%
		.[["meanGC"]],
	lengths = enframe(rownames(x), name = c()) %>%
		dplyr::rename(gene_id = value) %>% 
		left_join(gnBias) %>%
		.[["len"]],
	sizeFactors = x$samples$lib.size
)
x$offset <- gcCqn$glm.offset
```

```{r plotCQN, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible.*"}
par(mfrow = c(1, 2))
cqnplot(gcCqn, n = 1, xlab = "GC Content")
cqnplot(gcCqn, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

```{r, fig.cap = "*PCA after CQN showing similar patterns of separation as prior to this normalisation.*"}
logCPM <- gcCqn$y + gcCqn$offset
logCPM %>%
	t() %>%
	prcomp() %>%
	autoplot(data = x$samples, colour = "group", size = 4) +
	geom_text_repel(
		aes(label = sample, colour = group),
		show.legend = FALSE
	) +
	labs(colour = "Genotype")
```

## Likelihood ratio tests

This code below tests for differential expression using Likelihood ratio tests. 
Genes were considered DE if the FDR adjusted p-value was < 0.05. 

```{r}
glmCQN <- glmFit(x, design) 
topTableCQN <- glmCQN %>% 
	glmLRT(coef = "FMR1") %>%
	topTags(n = Inf, adjust.method = "fdr", sort.by = "p") %>% 
	.$table %>% 
	as_tibble() %>% 
	mutate(DE = FDR < 0.05)
```

## Comparison of MD Plots

```{r, fig.cap = "*Comparison of MD plots before and after CQN. The bias evident in A) the pre-CQN plots is no longer present in B) the post-CQN plot, suggesting GC and length bias were contributing factors.*"}
md2 <- topTableCQN %>% 
	ggplot(aes(x = logCPM, y = logFC)) +
	geom_point(aes(colour = DE), alpha = 0.5) +
	geom_smooth(se = FALSE) +
	scale_color_manual(values = c("grey50", "red")) +
	geom_label_repel(
		aes(label = external_gene_name, colour = DE), 
		data = . %>% dplyr::filter(DE == TRUE), 
		show.legend = FALSE
	)
plot_grid(
	md1 + 
		coord_cartesian(ylim = c(-3, 3)) + 
		theme(legend.position = "none"),
	md2 + 
		coord_cartesian(ylim = c(-3, 3)) + 
		theme(
			legend.position = c(1, 0), 
			legend.justification = c(1, 0),
			legend.background = element_rect(colour = "grey30", size = 1/4)
		),
	nrow = 1,
	labels = c("A", "B")
)
```


## Volcano plot

```{r}
topTableCQN %>% 
	ggplot(
		aes(x = logFC, y = -log10(PValue), colour = DE)
	) +
	geom_point(alpha = 0.5) +
	scale_color_manual(values = c("grey50", "red")) +
	geom_label_repel(
		aes(label = external_gene_name), 
		data = . %>% dplyr::filter(DE == TRUE), 
		show.legend = FALSE
	) +
	theme_bw()
```


A total of `r dplyr::filter(topTableCQN, DE == TRUE) %>% nrow()` genes were considered as significant as given in the table below.


```{r}
topTableCQN %>% 
	dplyr::filter(DE) %>%
	dplyr::select(
		ID = ensembl_gene_id,
		Gene = external_gene_name, 
		starts_with("log"),
		PValue, FDR, 
		Location = location, 
		Description = description
	) %>%
	datatable(
		caption = "The most highly ranked genes for differential expression.",
		rownames = FALSE,
		options = list(
			pageLength = 25
		)
	) %>%
	formatRound(
		columns = c("logFC", "logCPM"), 
		digits = 2
	) %>%
	formatSignif(
		columns = c("PValue", "FDR"),
		digits = 3
	)
```

```{r}
topTable_glm %>%
 	write.csv(here::here("results/DE_results_fmr1KO_KB.csv"))
```

# Visualisation of the DE genes

## Boxplots of Expression Values 

LogCPMs of the significantly DE genes are shown in the plot below. 

```{r, fig.cap = "*Boxplots of all DE genes with points overlaid*"}
logCPM %>%
	.[dplyr::filter(topTableCQN, DE == TRUE)$ensembl_gene_id,] %>% 
	as.data.frame() %>% 
	rownames_to_column("ensembl_gene_id") %>% 
	gather(key = "sample", value = "logCPM", colnames(x$counts)) %>% 
	left_join(x$samples, by = "sample") %>% 
	left_join(x$genes) %>% 
	ggplot(aes(x = group, y = logCPM, fill = group)) +
	geom_boxplot(alpha = 0.5) + 
	geom_jitter(width = 0.1, colour = "grey30") +
	facet_wrap(~external_gene_name, scales = "free_y", ncol = 4) +
	theme_bw() +
	theme(legend.position = "none") 
```


#  Enrichment Testing 

```{r}
pwf <- topTableCQN %>%
	left_join(gnBias, by = c("ensembl_gene_id" = "gene_id")) %>%
	with(
		nullp(
			DEgenes = structure(DE, names = ensembl_gene_id),
			bias.data = meanGC
		)
	)
```

## Defining Gene Sets

```{r}
ens2Entrez <- genes[rownames(x)] %>%
	mcols() %>%
	as.data.frame() %>%
	as_tibble() %>%
	dplyr::select(
		gene_id, 
		entrez_gene = entrezid
	) %>%
	unchop(entrez_gene) %>%
	dplyr::filter(!is.na(entrez_gene)) 
```


```{r}
geneByPos <- x$genes %>% 
	as_tibble() %>%
	separate(location, into = c("chr", "range", "strand"), sep = ":") %>% 
	split(f = .$ensembl_gene_id) %>% 
	lapply(extract2, "chr")
posByChr <- x$genes %>% 
	as_tibble() %>%
	separate(location, into = c("chr", "range", "strand"), sep = ":") %>% 
	split(f = .$chr) %>% 
	lapply(extract2, "ensembl_gene_id")
```

```{r}
hm <- msigdbr("Danio rerio", category = "H") %>%
	inner_join(ens2Entrez) %>%
	distinct(gs_name, gene_id, .keep_all = TRUE)
hmByGene <- hm %>%
  split(f = .$gene_id) %>%
  lapply(extract2, "gs_name")
hmByID <- hm %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")
```

```{r}
kg <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>%
	inner_join(ens2Entrez) %>%
	distinct(gs_name, gene_id, .keep_all = TRUE)
kgByGene <- kg %>%
  split(f = .$gene_id) %>%
  lapply(extract2, "gs_name")
kgByID <- kg %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")
```

## Testing Within the Set of DE Genes

### Chromosomal Position

```{r}
goseq(pwf, gene2cat = geneByPos) %>% 
	dplyr::filter(numDEInCat > 0) %>%
	as_tibble() %>%
	mutate(
		adjP = p.adjust(over_represented_pvalue, "bonferroni"),
		Expected = sum(topTableCQN$DE) * numInCat / nrow(x)
	) %>%
	dplyr::select(
		chromosome = category, Expected, Observed = numDEInCat, N = numInCat, PValue = over_represented_pvalue, adjP
	) %>%
	pander(
		caption = "Results for enrichment testing within the DE set of genes, based on chromosomal location."
	)
```

### Hallmark GenSets

```{r}
goseq(pwf, gene2cat = hmByGene) %>% 
	dplyr::filter(numDEInCat > 0) %>%
	as_tibble() %>%
	mutate(
		adjP = p.adjust(over_represented_pvalue, "bonferroni"),
		Expected = sum(topTableCQN$DE) * numInCat / nrow(x)
	) %>%
	dplyr::select(
		Category = category, Expected, Observed = numDEInCat, `GS Size` = numInCat, PValue = over_represented_pvalue, adjP
	) %>%
	pander(
		caption = "Results for enrichment testing for Hallmark Gene Sets within the DE set of genes",
		split.tables = Inf,
		justify = "lrrrrr"
	)
```

### KEGG Pathways

```{r}
goseq(pwf, gene2cat = kgByGene) %>% 
	dplyr::filter(numDEInCat > 0) %>%
	as_tibble() %>%
	mutate(
		adjP = p.adjust(over_represented_pvalue, "bonferroni"),
		Expected = sum(topTableCQN$DE) * numInCat / nrow(x)
	) %>%
	dplyr::select(
		Category = category, Expected, Observed = numDEInCat, `GS Size` = numInCat, PValue = over_represented_pvalue, adjP
	) %>%
	pander(
		caption = "Results for enrichment testing for KEGG Gene Sets within the DE set of genes",
		split.tables = Inf,
		justify = "lrrrrr"
	)
```



Since only 14 genes were found as DE, I will perform a GSEA on ranked lists (rather than overrepresentation in the DE genes) to see if any groups of genes within pre-defined gene sets are showing shifts towards up or downregulation as a group. I will look at the KEGG and HALLMARK gene sets available from MSigDB (https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp). These gene sets have human gene entrez identifiers. So I will convert these human entrez ids to zebrafish ensembl ids using a mapping file downloaded from bioMart (https://m.ensembl.org/biomart). Some genes in the KEGG pathway gene sets did not contain a zebrafish orthologue. Therefore, the gene sets sometimes contained < 10 genes. I filtered out any KEGG pathway gene sets with less than 10 genes as they would not be particularly informative. 

## Testing Using Ranked Lists

### FRY

`fry` from the `limma` package approximates the ROAST method. ROAST uses residual space rotation as a sort of continuous version of sample permutation. Like permutation tests, it protects against false positives caused by correlations between genes in the set (Wu et al, 2010). Using this method, I did not find any significantly altered gene sets in a directional context (FDR column), or mixed (so no directionality is taken into account (FDR.mixed column)). 

```{r}
fryRes <- logCPM %>%
	fry(
		index = c(hmByID, kgByID, posByChr), 
		design = design, 
		contrast = "FMR1", 
		sort = "mixed"
	) %>% 
	rownames_to_column("geneset") %>% 
	as_tibble() 
	
fryRes %>%
	arrange(PValue.Mixed) %>% 
	mutate(FDR.Mixed = p.adjust(PValue.Mixed, "fdr")) %>%
	dplyr::select(geneset, NGenes, PValue = PValue.Mixed, FDR = FDR.Mixed) %>%
	dplyr::slice(1:10) %>%
	pander(
		caption = paste(
			"The", nrow(.), "most highly ranked gene sets across Chromosomal, Hallmark and KEGG testing using Fry.",
			"Tests were non-directional (Mixed)."
		),
		split.tables = Inf,
		justify = "lrrr"
	)
```

# Enrichment testing on chromosome 14 genes. 

Fry showed that genes on chromosome 14 tend to be differentially expressed. To determine whether the top most altered genes from chromosome 14 are enriched in any KEGG or HALLMARK gene sets, I will performe `fgsea` using the chromosomal position gene sets on an undirectional ranked list (i.e. most DE -> to least DE). The `leading edge` fromt fgsea will give the genes which are the most dysregulated from chromosome 14. 

I then will perform an over-representation analysis (using the KEGG and HALLMARK gene sets) on the leading edge genes to determine whether they are have a particular biological function. 

```{r fgsea}
ranks <- topTableCQN %>% 
	mutate(rank = log10(1/PValue)) %>% 
	arrange(rank) %>% 
	dplyr::select(c("ensembl_gene_id", "rank")) %>% #only want the Pvalue with sign
	with(structure(rank, names = ensembl_gene_id)) %>% 
	rev()
```

```{r plotranks, fig.cap="Visualisation of ranked list"}
png("rankedlist.png", width = 10, height = 10, units = "cm", res = 600)
plot(ranks, 
	 xlab = "Genes", ylab = "-log10(PValue)")
dev.off()
```

The GSEA results indicate that the HALLMARK_E2F_TARGETS, HALLMARK_G2M_CHECKPOINT, KEGG_DNA_REPLICATION, chr14, and chr22 genes appear to be enriched at the start of the ranked list. 

```{r fgsea1}
# perform fgsea on all the gene sets. 
gsea_chr <- fgsea(pathways =  posByChr, 
	  stats = ranks, 
	  nproc = 1, 
	  nperm = 1e5
) %>% 
	as_tibble()

gsea_chr %>% 
	arrange(pval) %>% 
	mutate(padj = p.adjust(pval, "bonferroni")) %>% 
	dplyr::filter(padj < 0.05) %>% 
	dplyr::select(-leadingEdge) %>% 
	pander(
		caption = "Results for enrichment testing using fgsea based on chr position."
	)
	
```

The plot below shows the enrichment plot for chromosome 14. 

```{r, fig.cap = "*Enrichment plots of genes on chromosome 14 and 22*"}
# plot the enrichment
png("chr14_enrichment.png", width = 10, height = 10, units = "cm", res = 600)
plotEnrichment(pathway = posByChr$`14`,
			   stats = ranks)
dev.off()

png("chr22_enrichment.png", width = 10, height = 10, units = "cm", res = 600)
plotEnrichment(pathway = posByChr$`22`,
			   stats = ranks)
dev.off()
```

Next, I will take the leading edge for chromosome 14 (i.e. the genes in the ranked list from chromosome 14 which are before the peak on the enrichment plot above) and determine whether they are over-represented in any KEGG gene sets using the `kegga` function from the `limma` package. 

First, we perfomed over-representation analysis using `kegga()` of the leading edge subset against all detectable genes in the RNA-seq experiment. The KEGG gene sets for RNA polymerase and NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION were found to be signifcatly enriched (by FDR). 

```{r}
kegga(de = gsea_chr %>% dplyr::filter(pathway == "14") %>% 
	  	.$leadingEdge %>% .[[1]], 
	  geneid = rownames(x$counts), 
	  universe = rownames(x$counts),
	  gene.pathway = kgByID %>%
	  	unlist %>%
	  	as.data.frame() %>%
	  	rownames_to_column("pathway") %>%
	  	set_colnames(c("pathway", "gene_id")) %>%
	  	mutate(pathway = pathway %>% str_remove_all("[0-9]+$")) %>%
	  	dplyr::select(gene_id, pathway)
) %>% 
	topKEGG() %>% 
	as.data.frame() %>% 
	rownames_to_column("pathway") %>% 
	mutate(FDR = p.adjust(P.DE, "fdr"),
		   pbonf = p.adjust(P.DE, "bonferroni")) %>% 
	dplyr::select(pathway, raw.p = P.DE, FDR, pbonf, `No. in leadingedge` = DE, `No. in geneset` = N) %>%
	head(10) %>% 
	pander(caption = "Results of over-representation of genes in the leading edge of chromosome 14 against all genes detetcted in RNA-seq experiment")
```

As an alternate viewpoint, an over-representation analysis was performed on the leading edge genes from chr14 relative to all genes on chromosome 14. No significant gene sets were observed. 

```{r}
kegga(de = gsea_chr %>% dplyr::filter(pathway == "14") %>% 
	  	.$leadingEdge %>% .[[1]], 
	  geneid = rownames(x$counts), 
	  universe = x$genes %>% 
	  	dplyr::filter(grepl(x = location, pattern = "^14:")) %>% 
	  	.$ensembl_gene_id,
	  gene.pathway = kgByID %>%
	  	unlist %>%
	  	as.data.frame() %>%
	  	rownames_to_column("pathway") %>%
	  	set_colnames(c("pathway", "gene_id")) %>%
	  	mutate(pathway = pathway %>% str_remove_all("[0-9]+$")) %>%
	  	dplyr::select(gene_id, pathway)
) %>% 
	topKEGG() %>% 
	as.data.frame() %>% 
	rownames_to_column("pathway") %>% 
	mutate(FDR = p.adjust(P.DE, "fdr"),
		   pbonf = p.adjust(P.DE, "bonferroni")) %>% 
	dplyr::select(pathway, raw.p = P.DE, FDR, pbonf, `No. in leadingedge` = DE, `No. in geneset` = N) %>%
	head(10) %>% 
	pander(caption = "Results of over-representation of genes in the leading edge of chromosome 14 against all genes on chr 14")
```

# check for nacre contamination

It has come to our attention that some of the samples (i.e. pools of larvae) may have arisen from fish heterozygous for a *nacre* (w2) allele of the gene *mifta*, as well as the *fmr1* gene. 

In each pool (sample), the larvae which showed a *nacre* phenotype (i.e. no pigment cells) were removed from the analysis. However, these would be larvae which are *homozygous* for the nacre allele. Heterozygous *nacre* larvae could also be present in this sample. 

To investigate to what extent there are contaminiating nacre larvae in our RNA-seq samples, I will look at the number of reads which align to the nacre site on chromosome 6. 

```{r}
# the bam files are on th mounted drive on phoenix. 
bams <- list.files("/fast/users/a1211024/fmr1/03_alignedData/bams", pattern = ".bam")

# turn this off as we aligned to ensembl genome
options(ucscChromosomeNames=FALSE)

#make the reference sequence with the chromosome name as 6, not chr6
temp = DNAStringSet(c("6"=paste(as.vector(Drerio$chr6), length(Drerio$chr6), TRUE),collapse=""))

atrack_s2 <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873955_GSM4578112_fmr1_1_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
	name = "Alignment", max.height = 10, 
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"), 
	size = 10)

png("gviz/sample_s2.png", width = 2.5, height = 20, units = "cm", res = 1000)
plotTracks(list(atrack_s2), chromosome = "6",
		   from = 43429185, to = 43429186, main = "S2", sizes = 1)
dev.off()


atrack_s4 <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873956_GSM4578113_fmr1_2_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
	name = "Alignment", max_height = 10, 
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"))

png("gviz/sample_s4.png", width = 2.5, height = 20, units = "cm", res = 1000)
plotTracks(list(atrack_s4), chromosome = "6",
		   from = 43429185, to = 43429186, main = "S4")
dev.off()

atrack_s5 <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873957_GSM4578114_fmr1_3_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
	name = "Alignment",
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"), 
	size = 10)

plotTracks(list(genomeaxis, atrack_s5), chromosome = "6",
		   from = 43429185, to = 43429186, main = "S5")

atrack_s8 <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873958_GSM4578115_fmr1_4_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
	 name = "Alignment",
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"), 
	size = 10)

plotTracks(list(genomeaxis, atrack_s8), chromosome = "6",
		   from = 43429185, to = 43429186, main = "S8")

atrack_A <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873959_GSM4578116_WT_1_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
name = "Alignment",
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"), 
	size = 10)

plotTracks(list(atrack_A), chromosome = "6",
		   from = 43429185, to = 43429186, main = "A")


atrack_D <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873960_GSM4578117_WT_2_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
name = "Alignment",
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"), 
	size = 10)

plotTracks(list(atrack_D), chromosome = "6",
		   from = 43429185, to = 43429186, main = "D")

atrack_G <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873961_GSM4578118_WT_3_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
name = "Alignment",
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"))

plotTracks(list(atrack_G), chromosome = "6",
		   from = 43429185, to = 43429186, main = "G")

atrack_L <- AlignmentsTrack(
	"/fast/users/a1211024/fmr1/03_alignedData/bams/SRR11873962_GSM4578119_WT_4_Danio_rerio_RNA-Seq_Aligned.sortedByCoord.out.bam", 
	isPaired = TRUE, chromosome = "6", genome = "danRer11", 
name = "Alignment",
	referenceSequence = SequenceTrack(temp, 
									  name = "Ref", chromosome = "6"))

plotTracks(list(atrack_L), chromosome = "6",
		    from = 43429185, to = 43429186, main = "L")
```

```{r fig.cap = "correlation of the first four principal components with the proportion of reads aligning to the nacre allele."}
# table showing number of reads aligning to the W2 site according to samtools view
nacre <- tibble(sample = colnames(x), 
	   num_reads = c(39, 51, 55, 44, 47, 58, 65, 42), 
	   num_nacre_reads = c(1, 4, 16, 0.000001, 0.000001, 16, 0.000001, 3)) %>% 
	mutate(prop_nacre = num_nacre_reads*100/num_reads)

nacre
```

```{r, fig.cap="correlation of principal components with fmr1 genotype and proportion of nacre reads. The proportion of nacre reads has almost as much correlation as *fmr1* genotype with the firsr four PCs."}
# also look at PCA after log
logCPM %>% 
	t() %>% 
	prcomp %>% 
	.$x %>% 
	as.data.frame() %>% 
	rownames_to_column("sample") %>% 
	left_join(nacre %>% dplyr::select(sample, prop_nacre)) %>% 
	left_join(x$samples %>% dplyr::select(sample, c( fmr1_genotype = group), lib.size)) %>% 
	dplyr::select(-c(PC5, PC6, PC7, PC8)) %>% 
	mutate_at(vars(fmr1_genotype), as.integer) %>%
	column_to_rownames("sample") %>%
	cor() %>%
	corrplot::corrplot(
		diag = FALSE, 
		title = "correlation between the first four PCs of the cqn adjusted logCPMs, fmr1 genotype, library size, and proportion of reads aligning to nacre allele",
		type = "upper", 
		addCoef.col = "black"
  )
logCPM %>% 
	t() %>% 
	prcomp %>% 
	.$x %>% 
	as.data.frame() %>% 
	rownames_to_column("sample") %>% 
	left_join(nacre %>% dplyr::select(sample, prop_nacre)) %>% 
	left_join(x$samples %>% dplyr::select(sample, c( fmr1_genotype = group), lib.size)) %>% 
	pivot_longer(cols = starts_with("PC"), values_to = "PC_value", names_to = "Component") %>% 
	dplyr::filter(Component %in% c("PC1", "PC2", "PC3", "PC4")) %>% 
	ggplot(aes(x = PC_value, y = prop_nacre)) +
	geom_point(aes(colour = fmr1_genotype)) +
	theme_bw() +
	facet_wrap(~Component, scales = "free_x") +
	geom_label_repel(aes(label = sample)) +
	geom_smooth(method = "lm")

```

Some correlation is observed with the value of the first four principal components with the proportion of reads aligning to the nacre allele. However, it does not appear to be highly significantly due to te standard errors of the regression lines being quite large. 


## DGE analysis including nacre in the model

```{r}
# add nacre to x$samples
x$samples %<>% 
	left_join(nacre %>% 
			  	dplyr::select(sample, prop_nacre))
# design
design_nacre <- model.matrix(~group + prop_nacre, data = x$samples) %>%
  set_colnames(
    str_remove_all(colnames(.), "group")
  ) 

x_nac <- x %>% estimateDisp(design_nacre) %>% 
	glmFit(design_nacre) %>% 
	glmLRT(coef = "FMR1") %>%
	topTags(n = Inf, adjust.method = "fdr", sort.by = "p") %>% 
	.$table %>% 
	as_tibble() %>% 
	mutate(DE = FDR < 0.05)

```

When including proportion of nacre reads in the model, there are 189 DE genes (compared to 21 without). All 21 DE genes before including in the model are still found as DE. 

### comparison of MD plots

```{r}
x_nac %>% 
	ggplot(aes(x = logCPM, y = logFC)) +
	geom_point(aes(colour = DE), alpha = 0.5) +
	geom_smooth(se = FALSE) +
	geom_label_repel(
		aes(label = external_gene_name, colour = DE), 
		size = 2,
		data = . %>% dplyr::filter(DE == TRUE), 
		show.legend = FALSE) +
	scale_color_manual(values = c("grey50", "red")) +
	theme_bw()
```

## GOSEQ

I now will repeat the goseq enrichment testing to determine whether the results change at all. 

```{r}
pwf_nacre <- x_nac %>%
	left_join(gnBias, by = c("ensembl_gene_id" = "gene_id")) %>%
	with(
		nullp(
			DEgenes = structure(DE, names = ensembl_gene_id),
			bias.data = meanGC
		)
	)
```

```{r}
goseq(pwf_nacre, gene2cat = geneByPos) %>% 
	dplyr::filter(numDEInCat > 0) %>%
	as_tibble() %>%
	mutate(
		adjP = p.adjust(over_represented_pvalue, "bonferroni"),
		Expected = sum(x_nac$DE) * numInCat / nrow(x)
	) %>%
	dplyr::select(
		chromosome = category, Expected, Observed = numDEInCat, N = numInCat, PValue = over_represented_pvalue, adjP
	) %>%
	pander(
		caption = "Results for enrichment testing within the DE set of genes, based on chromosomal location."
	)
```

### Hallmark GenSets

```{r}
goseq(pwf_nacre, gene2cat = hmByGene) %>% 
	dplyr::filter(numDEInCat > 0) %>%
	as_tibble() %>%
	mutate(
		adjP = p.adjust(over_represented_pvalue, "bonferroni"),
		Expected = sum(x_nac$DE) * numInCat / nrow(x)
	) %>%
	dplyr::select(
		Category = category, Expected, Observed = numDEInCat, `GS Size` = numInCat, PValue = over_represented_pvalue, adjP
	) %>%
	pander(
		caption = "Results for enrichment testing for Hallmark Gene Sets within the DE set of genes",
		split.tables = Inf,
		justify = "lrrrrr"
	)
```

### KEGG Pathways

```{r}
goseq(pwf_nacre, gene2cat = kgByGene) %>% 
	dplyr::filter(numDEInCat > 0) %>%
	as_tibble() %>%
	mutate(
		adjP = p.adjust(over_represented_pvalue, "bonferroni"),
		Expected = sum(x_nac$DE) * numInCat / nrow(x)
	) %>%
	dplyr::select(
		Category = category, Expected, Observed = numDEInCat, `GS Size` = numInCat, PValue = over_represented_pvalue, adjP
	) %>%
	head
	pander(
		caption = "Results for enrichment testing for KEGG Gene Sets within the DE set of genes",
		split.tables = Inf,
		justify = "lrrrrr"
	)
```

## enrichment testing on entire list

```{r}
fry_res_nacre <- logCPM %>%
	fry(
		index = c(hmByID, kgByID, posByChr), 
		design = design_nacre, 
		contrast = "FMR1", 
		sort = "mixed"
	) %>% 
	rownames_to_column("geneset") %>% 
	as_tibble() 
	
fry_res_nacre %>%
	arrange(PValue.Mixed) %>% 
	mutate(FDR.Mixed = p.adjust(PValue.Mixed, "fdr")) %>%
	dplyr::select(geneset, NGenes, PValue = PValue.Mixed, FDR = FDR.Mixed) %>%
	dplyr::filter(FDR < 0.05) %>%
	pander(
		caption = paste(
			"The", nrow(.), "most highly ranked gene sets across Chromosomal, Hallmark and KEGG testing using Fry.",
			"Tests were non-directional (Mixed)."
		),
		split.tables = Inf,
		justify = "lrrr"
	)
```





# BLAST analysis for genetic compensation

This part was taken from Stephen and Melanie's `FMR1_Analysis.Rmd` file:

For each of the 3 DE genes, a brief analysis was performed using the Ensembl BLAST server to search all cDNAs within the zebrafish transcriptome, with parameters set for distant homology. 
For each gene the longest transcript was chosen as the query sequence.
Only results with an E-value < 1 were retained.

Note that the 3 DE genes were found from the analysis using `limma`. These genes were also detected using edgeR.  

```{r loadFMR1blast, echo=FALSE}
fmr1Blast <- read_csv(
	here::here("results/fmr1_blastn_results.csv"),
	col_types = "cciicccciiccnnc"
) %>%
	mutate(Length = str_extract(Length, "[0-9]+") %>% as.integer())
```

```{r fmr1BlastTable, echo=FALSE}
fmr1Blast %>% 
	dplyr::filter(`Gene hit` != "fmr1") %>%
	group_by(`Gene hit`) %>%
	summarise(`Total Matches` = n(),
			  `Longest Match` = max(Length)) %>%
	arrange(desc(`Longest Match`), desc(`Total Matches`)) %>%
	left_join(as.data.frame(genes), by = c(`Gene hit` = "symbol")) %>%
	dplyr::select(`Gene hit`, description, `Total Matches`, `Longest Match`) %>%
	distinct(`Gene hit`, .keep_all = TRUE) %>%
	mutate(description = gsub("([^\\[]+).+", "\\1", description)) %>%
	pander(
		caption = "Summary of BLAST matches for *fmr1*",
		justify = "llrr"
	)
```

Below is the volcano plot from edgeR with the blast hits for *fmr1* highlighted. 

```{r}
topTableCQN %>%
	mutate(`Blast Hit` = external_gene_name %in% fmr1Blast$`Gene hit`) %>% 
	arrange(`Blast Hit`) %>%
	ggplot(aes(logFC, -log10(PValue), label = external_gene_name, colour = `Blast Hit`)) +
	geom_point(alpha = 0.4) +
	geom_label_repel(data = . %>% dplyr::filter(`Blast Hit`),
					show.legend = FALSE, 
					size = 3) +
	scale_colour_manual(values = c("grey30", "blue")) +
	theme_bw() +
	theme(aspect.ratio = 1) +
	ggsave("blastRe_edgeR.png")
```

## BLAST Analysis (Other genes)

From *FMR1_Analysis.Rmd*:

Although not the subject of investigation, the genes detected as DE were analysed using BLAST as a reverse viewpoint on the experiment.
None of the genes contained matches to *fmr1*.

```{r loadSMTNLBlaast, echo=FALSE}
smtnl1Blast <- read_csv(
	here::here("results/smtnl1_blastn_results.csv"),
	col_types = "cciicccciiccnnc"
) %>%
	mutate(Length = str_extract(Length, "[0-9]+") %>% as.integer())
```

```{r smtnl1BlastTable, echo=FALSE}
smtnl1Blast %>% 
	dplyr::filter(`Gene hit` != "smtnl1") %>%
	group_by(`Gene hit`) %>%
	summarise(`Total Matches` = n(),
			  `Longest Match` = max(Length)) %>%
	arrange(desc(`Longest Match`), desc(`Total Matches`)) %>%
	left_join(as.data.frame(genes), by = c(`Gene hit` = "symbol")) %>%
	dplyr::select(`Gene hit`, description, `Total Matches`, `Longest Match`) %>%
	distinct(`Gene hit`, .keep_all = TRUE) %>%
	mutate(description = gsub("([^\\[]+).+", "\\1", description)) %>%
	pander(
		caption = "Summary of BLAST matches for *smtnl1*",
		justify = "llrr",
		split.tables = Inf
	)
```

```{r volcanoSMTNL1, echo=FALSE, fig.cap="Volcano plot showing closest genes from the BLAST search on *smtnl1*"}
topTableCQN %>%
	mutate(`Blast Hit` = external_gene_name %in% smtnl1Blast$`Gene hit`) %>%
	arrange(`Blast Hit`) %>%
	ggplot(aes(logFC, -log10(PValue), label = external_gene_name, colour = `Blast Hit`)) +
	geom_point(alpha = 0.4) +
	geom_label_repel(data = . %>% dplyr::filter(`Blast Hit`),
					show.legend = FALSE) +
	scale_colour_manual(values = c("grey50", "blue")) +
	theme_bw()
```

One gene (*nop58*) with a BLAST match to *smtnl* narrowly missed the threshold for being called DE.
However, the match was only 35bp.

```{r loadCU_Blast, echo=FALSE}
CU468164.1Blast <- read_csv(
	here::here("results/CU468164.1_blastn_results.csv"),
	col_types = "cciiccciicnnc"
) %>%
	mutate(Length = str_extract(Length, "[0-9]+") %>% as.integer())
```

```{r CU_BlastTable, echo=FALSE}
CU468164.1Blast %>% 
	dplyr::filter(`Gene hit` != "CU468164.1") %>%
	group_by(`Gene hit`) %>%
	summarise(`Total Matches` = n(),
			  `Longest Match` = max(Length)) %>%
	arrange(desc(`Longest Match`), desc(`Total Matches`)) %>%
	left_join(as.data.frame(genes), by = c(`Gene hit` = "symbol")) %>%
	dplyr::select(`Gene hit`, description, `Total Matches`, `Longest Match`) %>%
	distinct(`Gene hit`, .keep_all = TRUE) %>%
	mutate(description = gsub("([^\\[]+).+", "\\1", description)) %>%
	pander(
		caption = "Summary of BLAST matches for *CU468164.1*",
		justify = "llrr",
		split.tables = Inf
	)
```

```{r volcanoCU_, echo=FALSE, fig.cap="Volcano plot showing closest genes from the BLAST search on *smtnl1*"}
topTable_glm %>%
	mutate(`Blast Hit` = external_gene_name %in% c("CU468164.1", CU468164.1Blast$`Gene hit`)) %>%
	arrange(`Blast Hit`) %>%
	ggplot(aes(logFC, -log10(PValue), label = external_gene_name, colour = `Blast Hit`)) +
	geom_point(alpha = 0.4) +
	geom_label_repel(data = . %>% dplyr::filter(`Blast Hit`),
					show.legend = FALSE) +
	scale_colour_manual(values = c("grey50", "blue")) +
	theme_bw()
```

# Conclusion

From *FMR1_Analysis.Rmd*:
At this stage, no clear evidence can be found for genetic compensation in this dataset.
Alternative strategies were briefly investigated using pair-wise alignments between all DE genes, but no compelling results were found. As no specific evidence exists for 3'UTR, 5'UTR or other non-coding regions in compensation, no analysis was conducted on these elements.

# Session info
```{r}
sessionInfo()
```


